Add hide channel option directly to More Options menu

MitchelPaulin opened the pull request:
# Add hide channel option directly to More Options menu

<!-- Thanks for sending a pull request! Make sure to follow the contributing guidelines. -->
<!-- Important note, we may remove your pull request if you do not use this provided PR template correctly. -->

## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue
<!-- Please link the issue your pull request is referring to. -->
<!-- If this pull request fully resolves the relevant issue, put "closes" before the issue number. -->
<!-- Example: "closes #123456". -->

closes https://github.com/FreeTubeApp/FreeTube/issues/3051

## Description

Piggybacks off the feature that was implemented in https://github.com/FreeTubeApp/FreeTube/pull/2849 to add an option directly to the "More Options" menu to hide a channel. Options will show up in the "Distraction Settings" afterwards and can be removed from there. 


## Screenshots <!-- If appropriate -->
<!-- Please add before and after screenshots if there is a visible change. -->

![image](https://github.com/FreeTubeApp/FreeTube/assets/30714020/b25e7148-fc41-457e-8089-390c4b101cc8)


## Testing <!-- for code that is not small enough to be easily understandable -->
<!-- Has this pull request been tested? -->
<!-- Please describe shortly how you tested it. -->
<!-- Are there any ramifications remaining? -->

Selected the option in the context menu, the video is immediately filtered and a toast is thrown up to confirm it was hidden. Then I checked the "Distraction Settings" and confirmed the channel is there. 

## Desktop
<!-- Please complete the following information-->
- **OS:** Pop_OS
- **OS Version:** 22.04 LTS
- **FreeTube version:** 0.19.1

## Additional Info

I decided not to go with "Hide Recommendation" like mentioned in the original issue as there really isn't recommendations, just trending.

PikachuEXE commented:
I think the new option should be added to the end with a separator
Also it might be better to add a confirmation modal/something to avoid misclick

jasonhenriquez commented:
Good work @MitchelPaulin!

On top of @PikachuEXE's suggestions, could you also add an "Unhide Channel" option for when the channel is hidden? That may sound silly, but hidden channel videos can still show up in other places like Subscriptions, the Channel page, and Playlists.

MitchelPaulin commented:
Moved it into its own divider at the bottom and added the ability to show hidden channels, good suggestions. 

Not sure how I feel about the confirmation popup, I feel like it might just get in the way more than anything. But if its a blocker I can add it.

jasonhenriquez commented:
I don't think it's strictly needed because you can always see and remove those channels from your Distraction Free Settings. Especially given that there's a toast letting you know what you did if you accidentally clicked it. @PikachuEXE Second thoughts?

PikachuEXE reviewed (approved):
Hmm? Forgot to approve after review

absidue commented on src/renderer/components/ft-list-video/ft-list-video.js:
```diff
@@ -621,12 +644,28 @@ export default defineComponent({
       showToast(this.$t('Video.Video has been removed from your saved list'))
     },
 
+    hideChannel: function(channelName) {
+      const hiddenChannels = JSON.parse(this.$store.getters.getChannelsHidden)
+      hiddenChannels.push(channelName)
+      this.updateChannelsHidden(JSON.stringify(hiddenChannels))
+
+      showToast(this.$t('Channel Hidden', { channel: channelName }))
+    },
```
Please change this pull request to use the channel ID, not the name. Blocking by name is unreliable as the channel owner can change the channel name, so you would have to block it again when that happens, but the ID never changes, so it will continue to work even when the name changes.

MitchelPaulin replied:
I went with the name since the Ids were very unreadable, if #4230 merges though that's no longer a concern, I can switch it Ids

PikachuEXE commented:
I will wait for https://github.com/FreeTubeApp/FreeTube/pull/4230 before merging this

MitchelPaulin commented:
@PikachuEXE #4230 was merged, can this be merged now?

PikachuEXE reviewed (approved):
Oops sorry I forgot
No wonder I feel like something is missing when testing https://github.com/FreeTubeApp/FreeTube/pull/4230 ...

PikachuEXE commented:
@absidue Please re-review 😅

efb4f5ff-1298-471a-8973-3d47447115dc commented on static/locales/en-US.yaml:
```diff
@@ -920,6 +922,8 @@ Starting download: 'Starting download of "{videoTitle}"'
 Downloading failed: 'There was an issue downloading "{videoTitle}"'
 Screenshot Success: Saved screenshot as "{filePath}"
 Screenshot Error: Screenshot failed. {error}
+Channel Hidden: Channel {channel} added to filter
```
I think this messaging can be better now that #4202 is here

I think `{channel} added to Channel filter` is more explicit

efb4f5ff-1298-471a-8973-3d47447115dc commented on static/locales/en-US.yaml:
```diff
@@ -920,6 +922,8 @@ Starting download: 'Starting download of "{videoTitle}"'
 Screenshot Success: Saved screenshot as "{filePath}"
 Screenshot Error: Screenshot failed. {error}
+Channel Hidden: Channel {channel} added to filter
+Channel Unhidden: Channel {channel} removed from filter
```
same here

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
Maybe this can be made more explicit

efb4f5ff-1298-471a-8973-3d47447115dc commented on src/renderer/components/ft-list-video/ft-list-video.js:
```diff
@@ -228,8 +228,25 @@ export default defineComponent({
           )
+
+          const hiddenChannels = JSON.parse(this.$store.getters.getChannelsHidden)
+          const channelShouldBeHidden = hiddenChannels.some(c => c === this.channelName || c === this.channelId)
```
AFAIK channelName is not a thing anymore

MitchelPaulin replied:
Would we not need the channel name checked here to support anyone that has legacy blocks via channel name?

efb4f5ff-1298-471a-8973-3d47447115dc replied:
i could be missing something but

If im looking at the conversations in https://github.com/FreeTubeApp/FreeTube/pull/4230 i think everything is based on channel ID now

MitchelPaulin replied:
I was just thinking of what would happen to a user who has blocked a channel by name then updates freetube

MitchelPaulin replied:
Ah ok, scrolling through the other PR it does look like we are fine just not respecting legacy blocks https://github.com/FreeTubeApp/FreeTube/pull/4230#issuecomment-1783769677

In that case I can remove the channel name check

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
In the latest nightly this isnt a thing

efb4f5ff-1298-471a-8973-3d47447115dc commented:
I see `channelName` being used in more places do they also need to be removed?

MitchelPaulin commented:
> I see `channelName` being used in more places do they also need to be removed?

Yes, missed a spot, the rest of the uses are for the toast message though, so those need to stay

absidue commented:
Looks like an error caused by invalid YAML. The newly added strings with the curly bracket at the start probably need to be quoted.

absidue commented:
Thank you for your contribution! sorry for the long back-and-forwards on this.

PikachuEXE reviewed (approved):
I tested after merging `development` into this branch otherwise it might not work properly
All seems fine

BarbzYHOOL commented:
cool
